exercises/src/main/scala/fpinscala/laziness/Stream.scala

```diff
@@ -17,20 +27,95 @@ trait Stream[+A] {
+    case Cons(_, t) if n >= 1 => t().drop(n - 1)
+    case _ => this
+  }
+
+  //5.3
+  def takeWhile(p: A => Boolean): Stream[A] = this match {
+    case Cons(h, t) if p(h()) => cons(h(), t().takeWhile(p))
+    case _ => empty
+  }
 
-  def takeWhile(p: A => Boolean): Stream[A] = ???
+  //5.4
+  def forAll(p: A => Boolean): Boolean = foldRight(true)((a, b) => p(a) && b )
 
-  def forAll(p: A => Boolean): Boolean = ???
+  //5.5 takeWhile with foldRight
+  def takeWhile2(p: A => Boolean): Stream[A] = foldRight(Empty: Stream[A])((a, b) => if (p(a)) cons(a, b) else Empty)
 
-  def headOption: Option[A] = ???
+  //5.6 headOption with foldRight
+  def headOption: Option[A] = foldRight(None: Option[A])((a, _) => Some(a))
 
   // 5.7 map, filter, append, flatmap using foldRight. Part of the exercise is
   // writing your own function signatures.
+  def map[B](f: A => B): Stream[B] = foldRight(Empty: Stream[B])((a, b) => cons(f(a), b))
+
+  def filter(f: A => Boolean): Stream[A] = foldRight(Empty: Stream[A])((a, b) => if (f(a)) cons(a, b) else b)
+
+  def append[B >: A](x: => Stream[B]): Stream[B] = foldRight(x)((a, b) => cons(a, b))
+
+  def flatMap[B](f: A => Stream[B]): Stream[B] = foldRight(Empty: Stream[B])((a, b) => f(a) append b)
+
+  //5.13 map, take, takeWhile, zipWith, zipAll using unfold
+  def map2[B](f: A => B): Stream[B] = unfold(this) {
+    case Cons(h, t) => Some(f(h()), t())
+    case Empty => None
+  }
+
+  def take2(n: Int): Stream[A] = unfold((this, n)) {
+    case (Cons(h, t), x) if x > 0 => Some(h(), (t(), x - 1))
+    case _ => None
+  }
+
+  def takeWhile3(p: A => Boolean): Stream[A] = unfold(this) {
+    case Cons(h, t) if p(h()) => Some(h(), t())
+    case _ => None
+  }
+
+  def zipWith[B >: A](s: Stream[B])(f: (B, B) => B): Stream[B] = unfold((this, s)) {
+    case (Cons(h1, t1), Cons(h2, t2)) => Some(f(h1(), h2()), (t1(), t2()))
+    case _ => None
+  }
+
+  def zipAll[B](s: Stream[B]): Stream[(Option[A], Option[B])] = unfold((this, s)) {
+    case (Cons(h1, t1), Cons(h2, t2)) => Some((Some(h1()), Some(h2())), (t1(), t2()))
+    case (Cons(h, t), Empty) => Some((Some(h()), None), (t(), Empty))
+    case (Empty, Cons(h, t)) => Some((None, Some(h())), (Empty, t()))
+    case _ => None
+  }
+
+  def zipAllWith[B >: A](s: Stream[B])(f: (B, B) => B): Stream[B] = unfold((this, s)) {
+    case (Cons(h1, t1), Cons(h2, t2)) => Some(f(h1(), h2()), (t1(), t2()))
+    case (Cons(h, t), Empty) => Some(h(), (t(), Empty))
+    case (Empty, Cons(h, t)) => Some(h(), (Empty, t()))
+    case _ => None
+  }
+
+  //5.14
+  def startsWith[B](s: Stream[B]): Boolean = zipAll(s).forAll(x => x match {
+    case (Some(a), Some(b)) => a == b
+    case (_, None) => true
+    case _ => false
+  })
+
+  def startsWith2[B](s: Stream[B]): Boolean = zipAll(s).takeWhile(!_._2.isEmpty).forAll(x => x._1 == x._2)
+
+  //5.15
+  def tails: Stream[Stream[A]] = unfold (this) {
+    case Cons(h, t) => Some(Cons(h, t), t())
```

Comment:
Do you need to evaluate `t()` here? Can you make it a bit more lazy?

Reply:
I guess I should have used the lazy constructor cons

Reply:
That would have work, but you can also use `drop(1)` instead, quite often it's the safer alternative to get the `tail`.

```
case Empty => None
case s => Some(s, s.drop(1))
```

Reply:
Actually, `cons` wouldn't change anything here. And using `drop` might have the same degree of laziness on second though. It's just a bit shorter, might behave the same, I would have to look at the implementation of drop again to confirm.

Reply:
Yeah I was wrong, using `drop` would have the same degree of laziness here, sorry about that. Your original implementation is also better than the "correction" you did after my comment lead you in the wrong direction.